File: WebApplication/src/nomenclatures/filter-dtos/institution-filter.dto.ts
import { NomenclatureHierarchyFilterDto } from "src/nomenclatures/filter-dtos/nomenclature-hierarchy-filter.dto";
import { InstitutionDto } from "../dtos/institutions/institution.dto";
import { DistrictDto } from "src/nomenclatures/dtos/settlements/district.dto";
import { MunicipalityDto } from "src/nomenclatures/dtos/settlements/municipality.dto";
import { SettlementDto } from "src/nomenclatures/dtos/settlements/settlement.dto";
import { OrganizationType } from "../enums/organization-type.enum";
import { OwnershipType } from "../enums/ownership-type.enum";

export class InstitutionFilterDto extends NomenclatureHierarchyFilterDto<InstitutionDto> {
    uic: string;

    districtId: number;
    district: DistrictDto;
    municipalityId: number;
    municipality: MunicipalityDto;
    settlementId: number;
    settlement: SettlementDto;

    organizationType: OrganizationType;
    ownershipType: OwnershipType;
}

File: Sc.Models/FilterDtos/Nomenclatures/Institutions/InstitutionFilterDto.cs
using Sc.Models.Entities.Nomenclatures.Institutions;
using Sc.Models.Enums.Institutions;
using Sc.Models.FilterDtos.Base.Nomenclatures;

namespace Sc.Models.FilterDtos.Nomenclatures.Institutions
{
    public class InstitutionFilterDto : NomenclatureHierarchyFilterDto<Institution>
    {
        public string Uic { get; set; }
        public int? DistrictId { get; set; }
        public int? MunicipalityId { get; set; }
        public int? SettlementId { get; set; }
        public OrganizationType? OrganizationType { get; set; }
        public List<OrganizationType> OrganizationTypes { get; set; } = new List<OrganizationType>();
        public OwnershipType? OwnershipType { get; set; }

        public override IQueryable<Institution> WhereBuilder(IQueryable<Institution> query)
        {
            if (!string.IsNullOrWhiteSpace(Uic))
            {
                var uic = Uic.Trim().ToLower();
                query = query.Where(e => e.Uic.Trim().ToLower() == uic);
            }

            if (DistrictId.HasValue)
            {
                query = query.Where(e => e.DistrictId == DistrictId);
            }

            if (MunicipalityId.HasValue)
            {
                query = query.Where(e => e.MunicipalityId == MunicipalityId);
            }

            if (SettlementId.HasValue)
            {
                query = query.Where(e => e.SettlementId == SettlementId);
            }

            if (OrganizationType.HasValue)
            {
                query = query.Where(e => e.OrganizationType == OrganizationType);
            }

            if (OrganizationTypes != null && OrganizationTypes.Any())
            {
                query = query.Where(e => OrganizationTypes.Contains(e.OrganizationType.Value));
            }

            if (OwnershipType.HasValue)
            {
                query = query.Where(e => e.OwnershipType == OwnershipType);
            }

            return base.WhereBuilder(query);
        }
    }
}
